test/sample_test.yaml

```diff
@@ -145,6 +145,8 @@ spec:
           mirrorVolumeMounts: true
 
     - name: run-sample-tests
+      retryStrategy:
+        limit: 10
```

Comment:
@numerology How much time does it take to run sample tests for once?
I've got two concerns:
1. is this retry too high level? each retry will run all samples again, is that right?
2. what if one of the sample is indeed flaky? I think we should fix the sample because it also affects users, this will hide that. e.g. We used to see flakiness for workload identity, that was a real GKE bug. I think we'd want to catch that.

Therefore, I think we should only add retry for very explicit items that can fail (like image building, module fetching, license fetching....), also retries are more effective (increases success rate more) when it is added in individual flakiness points.

Reply:
> is this retry too high level? each retry will run all samples again, is that right?

No. When argo retry combines with withitem (aka parfor loop), each individual retry will be triggered separately. For example, [this](https://gist.github.com/numerology/d9f7e939476e9a5d61b094aade0a1701) pipeline def generates [this](https://7e84d722f4296aac-dot-us-central2.pipelines.googleusercontent.com/#/runs/details/49f2967d-003c-4aa5-a164-7790c5b4a25c) execution result. As you can see, different iterations fail different times.

> what if one of the sample is indeed flaky?
That's a good point. How about in postsubmit tests we disable retry (or equivalently set this to 1). This way, we can capture potential **real** flakiness while keep a healthy eng velocity.

> I think we should only add retry for very explicit items that can fail (like image building, module fetching, license fetching.

SGTM to also add retry to parts other than actual sample execution. BTW, I noticed we explicitly set retry == 1 [here](https://github.com/kubeflow/pipelines/blob/7b55c2b8101ec2d08b5c4c2f3a0d490d3ff8c0ee/test/sample_test.yaml#L111). Do you remember why we did that? Is there any idempotency issue?

Reply:
> > is this retry too high level? each retry will run all samples again, is that right?
> 
> No. When argo retry combines with withitem (aka parfor loop), each individual retry will be triggered separately. For example, [this](https://gist.github.com/numerology/d9f7e939476e9a5d61b094aade0a1701) pipeline def generates [this](https://7e84d722f4296aac-dot-us-central2.pipelines.googleusercontent.com/#/runs/details/49f2967d-003c-4aa5-a164-7790c5b4a25c) execution result. As you can see, different iterations fail different times.

Thanks for clarification! SGTM
 
> > what if one of the sample is indeed flaky?
> 
> That's a good point. How about in postsubmit tests we disable retry (or equivalently set this to 1). This way, we can capture potential **real** flakiness while keep a healthy eng velocity.

Good idea!
So changes needed:
* Run **full** sample test in post submit without retry
* I'd suggest adding retry == 1 or 2 here instead of 10, so we don't miss high flakiness from presubmit.

> > I think we should only add retry for very explicit items that can fail (like image building, module fetching, license fetching.
> 
> SGTM to also add retry to parts other than actual sample execution. BTW, I noticed we explicitly set retry == 1 [here](https://github.com/kubeflow/pipelines/blob/7b55c2b8101ec2d08b5c4c2f3a0d490d3ff8c0ee/test/sample_test.yaml#L111). Do you remember why we did that? Is there any idempotency issue?

It was initially added when workload identity was flaky, but it's good to retry generally because image building is always idempotent ... and flaky.

Reply:
Unfortunately, it seems like we cannot paramemeterize retryStrategy.limit as other argo parameters.I'll need to revert that commit.

Reply:
@numerology Shall we stick with retry limit = 2 first then? I think that should fix most flakiness now

Reply:
Definitely. (Sorry I forgot to update the e2e tests previously)